src/main/java/ch/njol/skript/registrations/SectionBuilder.java

```diff
@@ -0,0 +1,248 @@
+	}
+
+	/**
+	 * Defines the patterns for this section
+	 */
+	public SectionBuilder patterns(String... patterns) {
+		this.patterns = patterns;
+		return this;
+	}
+
+	/**
+	 * Defines the sections that this section is allowed to contain,
+	 * if not called then everything will be allowed (including {@link ch.njol.skript.lang.TriggerItem}).
+	 */
+	@SafeVarargs
+	public final SectionBuilder children(Class<? extends TriggerItem>... children) {
+		this.children = children;
+		return this;
+	}
+
+	/**
+	 * Defines the sections that this section is allowed to be inside as defined
+	 * by {@link SectionBuilder#placement(Section.Placement)}, if not called then
+	 * this section can be used anywhere.
+	 *
+	 * @throws IllegalArgumentException if {@code parents} is null or empty
+	 */
+	@SafeVarargs
+	public final SectionBuilder parents(Class<? extends TriggerSection>... parents) throws IllegalArgumentException {
+		this.parents = parents;
+		return placement(Section.Placement.DIRECT_CHILD);
+	}
+
+	/**
+	 * Defines how this section must be placed relative to it's
+	 * parents.
+	 *
+	 * @throws IllegalArgumentException if parents haven't been set to a
+	 * non-null, non-empty array via {@link SectionBuilder#parents(Class[])}
+	 * or if you attempt to use an invalid placement for the current parents.
+	 */
+	public SectionBuilder placement(Section.Placement placement) throws IllegalArgumentException {
+		if (ArrayUtils.isEmpty(parents)) {
+			throw new IllegalArgumentException("A section must have parents to use a placement");
+		}
+
+		if (placement != Section.Placement.CHILD && placement != Section.Placement.DIRECT_CHILD) {
+			throw new IllegalArgumentException("This placement is not allowed when a section has parents");
+		}
+
+		this.placement = placement;
+
+		return this;
+	}
+
+	/**
+	 * Sets the events that the code within this section will be parsed inside.
+	 *
+	 * Basically, this just sets {@link ch.njol.skript.ScriptLoader#currentEvents}
+	 * to {@code events} before the inner code is parsed.
+	 */
+	@SafeVarargs
+	public final SectionBuilder innerEvents(Class<? extends Event>... events) {
+		this.innerEvents = events;
+		return this;
+	}
+
+	/**
+	 * Defines the error printed when the section is placed somewhere
+	 * it shouldn't be as defined by {@link SectionBuilder#parents(Class[])}
+	 * and the allowed children of another method.
+	 */
+	public SectionBuilder illegalPlacementError(String error) {
+		illegalParentError(error);
+		return illegalChildError(error);
+	}
+
+	public SectionBuilder illegalParentError(String error) {
+		this.illegalParentError = error;
+		return this;
+	}
+
+	public SectionBuilder illegalChildError(String error) {
+		this.illegalChildError = error;
+		return this;
+	}
+
+
+	/**
+	 * Registers this section.
+	 *
+	 * @throws IllegalStateException if required info about this section is missing
+	 */
+	@SuppressWarnings("unchecked")
+	public void register() throws IllegalStateException {
+		if (name == null) {
```

Comment:
i need to check if registration is allowed, also it might make sense to use a skript api exception instead